crates/provider-sqldb-postgres/src/lib.rs

```diff
@@ -0,0 +1,514 @@
+            if connections.get(token).is_some() {
+                return Ok(token.into());
+            }
+        }
+
+        // Build the new connection pool
+        let runtime = Some(deadpool_postgres::Runtime::Tokio1);
+        let tls_required = create_opts.tls_required;
+        let cfg: deadpool_postgres::Config = create_opts.try_into()?;
+        let pool = if tls_required {
+            create_tls_pool(cfg, runtime)
+        } else {
+            cfg.create_pool(runtime, tokio_postgres::NoTls)
+                .context("failed to create non-TLS postgres pool")
+        }?;
+
+        // Save the newly created connection to the pool
+        let mut connections = self.connections.write().await;
+        connections.insert(token.into(), pool);
+        Ok(token.into())
+    }
+
+    /// Perform a query
+    async fn do_query(
+        &self,
+        connection_token: &str,
+        query: &str,
+        params: Vec<PgValue>,
+    ) -> Result<Vec<ResultRow>, QueryError> {
+        let connections = self.connections.read().await;
+        let pool = connections.get(connection_token).ok_or_else(|| {
+            QueryError::Unexpected(format!(
+                "missing connection pool for token [{connection_token}] while querying"
+            ))
+        })?;
+
+        let client = pool.get().await.map_err(|e| {
+            QueryError::Unexpected(format!("failed to build client from pool: {e}"))
+        })?;
+
+        let rows = client
+            .query_raw(query, params)
+            .await
+            .map_err(|e| QueryError::Unexpected(format!("failed to perform query: {e}")))?;
+
+        // todo(fix): once async stream support is available & in contract
+        // replace this with a mapped stream
+        rows.map_ok(into_result_row)
+            .try_collect::<Vec<_>>()
+            .await
+            .map_err(|e| QueryError::Unexpected(format!("failed to evaluate full row: {e}")))
+    }
+
+    /// Prepare a statement
+    async fn do_statement_prepare(
+        &self,
+        connection_token: &str,
+        query: &str,
+    ) -> Result<PreparedStatementToken, StatementPrepareError> {
+        let connections = self.connections.read().await;
+        let pool = connections.get(connection_token).ok_or_else(|| {
+            StatementPrepareError::Unexpected(format!(
+                "failed to find connection pool for token [{connection_token}]"
+            ))
+        })?;
+
+        let client = pool.get().await.map_err(|e| {
+            StatementPrepareError::Unexpected(format!("failed to build client from pool: {e}"))
+        })?;
+
+        let statement = client.prepare(query).await.map_err(|e| {
+            StatementPrepareError::Unexpected(format!("failed to prepare query: {e}"))
+        })?;
+
+        let statement_token = format!("prepared-statement-{}", Ulid::new().to_string());
+
+        let mut prepared_statements = self.prepared_statements.write().await;
```

Comment:
Do we need to have something that cleans up prepared statements or should they continuously grow?

Reply:
Prepared statements live for the lifetime of the connection that they were made on -- which is only when a link is dropped or shutdown, what were you thinking  here in terms of clean up?

Reply:
The thing I'm trying to avoid is let's say a component makes a bunch of prepared statements but then only uses them once. If there are 200 components connected to this provider, all with several hundred prepared statements that are only used once, is that going to take up a bunch of memory? This could absolutely not be a problem, but just wanted to call it out

Reply:
So it's definitely not going to take up much memory... compared to the actual data being queried -- and also note that the prepared statement is more importantly *on the DB side*.

I don't think this is a problem we need to worry about first -- connection pooling params would be more important to start